kernel/kernel-api/src/main/java/io/delta/kernel/coordinatedcommits/AbstractCommitCoordinatorBuilder.java

```diff
@@ -0,0 +1,133 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.coordinatedcommits;
+
+import io.delta.kernel.annotation.Evolving;
+import io.delta.kernel.config.ConfigurationProvider;
+import io.delta.kernel.internal.DeltaErrors;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+/**
+ * A builder class to create a {@link CommitCoordinatorClient} (CCC).
+ *
+ * <p>Subclasses of this class must provide a no-argument constructor.
+ *
+ * <p>Note: These builder classes are optional. The {@link
+ * io.delta.kernel.engine.Engine#getCommitCoordinatorClient} API does not prescribe how to create
+ * the underlying CCC and does not require a builder.
+ *
+ * <p>The benefit of implementing a builder for your CCC is that your Engine may then invoke {@link
+ * #buildCommitCoordinatorClient} to (1) instantiate your builder, and then (2) build a new CCC via
+ * reflection.
+ *
+ * <p>In short, this builder provides the ability for users to specify at runtime which builder
+ * implementation (e.g. x.y.z.FooCCBuilder) to use for a given commit coordinator name (e.g. "foo"),
+ * without imposing any restrictions on how to construct CCCs (i.e. these builders are optional).
+ *
+ * @since 3.3.0
+ */
+@Evolving
+public abstract class AbstractCommitCoordinatorBuilder {
+
+  ///////////////////////////
+  // Static Helper Methods //
+  ///////////////////////////
+
+  /**
+   * Builds the underlying {@link CommitCoordinatorClient} associated with the given commit
+   * coordinator name.
+   *
+   * <ul>
+   *   <li>Determines the specific builder configuration lookup key.
+   *   <li>Grabs the corresponding builder className for that key from the provided sessionConfig.
+   *   <li>Instantiates a new instance of that {@link AbstractCommitCoordinatorBuilder}.
+   *   <li>Invokes the builder's build method, passing along the sessionConfig and commit
+   *       coordinator config map.
+   * </ul>
+   *
+   * @param commitCoordinatorName the commit coordinator name
+   * @param sessionConfig The session-level configuration that may represent environment-specific
+   *     configurations such as {@code HadoopConf} or {@code SparkConf}. This sessionConfig is used
+   *     to look up the right builder className to instantiate for the given commit coordinator
+   *     name. It can also be used by builders to lookup per-session configuration values unique to
+   *     that builder.
+   * @param commitCoordinatorConf the parsed value of the Delta table property {@link
+   *     io.delta.kernel.internal.TableConfig#COORDINATED_COMMITS_TABLE_CONF} and represents the
+   *     configuration properties for describing the Delta table to the commit-coordinator.
+   * @return the {@link CommitCoordinatorClient} corresponding to the given commit coordinator name
+   */
+  public static CommitCoordinatorClient buildCommitCoordinatorClient(
```

Comment:
If I understand correctly, the dev experience for using this method will be:
```
CommitCoordinatorClient client = AbstractCommitCoordinatorBuilder
                .buildCommitCoordinatorClient(
                    "dynamodb",
                    sessionConfig,
                    coordinatorConf
                );
```
Somehow, it doesn't feel intuitive that this code will look up the builder in the conf and do stuff

Reply:
How's this method different than `CommitCoordinatorProvider.getBuilder(... /* load builder dynamically from conf */).build(...)`? And why do you prefer this one? 🤔

Reply:
https://github.com/delta-io/delta/blob/master/kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/coordinatedcommits/CommitCoordinatorProvider.java

is the old code that will be deleted when this is refactored. it uses HadoopConf and io.delta.storage APIs

Reply:
> Somehow, it doesn't feel intuitive that this code will look up the builder in the conf and do stuff

Really? I quite like it. Feel free to let me know what doesn't feel right about this. This is what we agreed on in the sync on Tuesday Oct 22 @ 1:00pm

Reply:
Further, this API feels _very_ similar to the previous API of

```
CommitCoordinatorClient client = CommitCoordinatorProvider
  .getCommitCoordinatorClient(
    hadoopConf,
    name,
    commitCoordinatorConf);
```

Reply:
I was hoping we could prevent patterns like the following by moving `buildCommitCoordinatorClient` to a separate `DefaultCommitCoordinatorClientBuilder` that `DynamoDBCCClientBuilder` doesn't need to extend:
```
CommitCoordinatorClient client = DynamoDBCCClientBuilder.buildCommitCoordinatorClient(
    "dynamodb", 
    config, 
    coordConf);
```